### PR TITLE
feat: add support for kafka transport

### PR DESCRIPTION
**Pull request by lance**
This commit extends the `message` package to include Kafka transport. The tests included here were almost entirely copied verbatim from `message_tests.ts` which tests the `HTTP` message binding. The primary change to these tests was the use of the `Kafka` binding instead of the `HTTP` binding. It would be nice, I think, to have generic tests that can test any new transport layer by just taking the binding as a parameter. But that's not what I've done here. :) I've mostly just copied and pasted.

There is one test that is pending as it's not passing yet, and I figured it was a bit of an edge case and could be addressed later.

```
      - Converts base64 encoded data to binary when deserializing structured messages
```

Note that only JSON encoding is supported.

Additionally, some of the type information has changed across the project to more accurately reflect the type of `Message` with `Message<T>`.

Related: https://github.com/cloudevents/sdk-javascript/issues/390

### CODE REVIEW

File: src/message/index.ts [hunk 1]
```diff
@@ -41,9 +41,9 @@ export interface Headers extends IncomingHttpHeaders {
  * @property {@linkcode Headers} `headers` - the headers for the event Message
  * @property string `body` - the body of the event Message
```
Comment (grant) [inline src/message/index.ts:42]: Looks like these docs should be updated with the code change.

Maybe other docs too.

File: src/message/index.ts [hunk 2]
```diff
@@ -41,9 +41,9 @@ export interface Headers extends IncomingHttpHeaders {
...
  */
-export interface Message {
+export interface Message<T> {
```
Comment (grant) [inline src/message/index.ts:44]: I believe all generic types require a default value to not be a breaking changes.

```ts
export interface Message<T = string> {
    headers: object;
    body: T | Buffer | unknown;
}
```

If this is a breaking change, please change the conventional commit type to `feat!`.

File: src/message/kafka/headers.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+ Copyright 2021 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+import { CloudEvent, CONSTANTS, Headers } from "../..";
+
+/**
+ * The set of CloudEvent headers that may exist on a Kafka message
+ */
+export const KAFKA_CE_HEADERS: { [key: string]: string } = Object.freeze({
```
Comment (grant) [inline src/message/kafka/headers.ts:11]: I believe the type could be stronger or be omitted, specifically the `key: string` type.

```ts
Readonly<{
    ID: string;
    TYPE: string;
    SOURCE: string;
    SPEC_VERSION: string;
    TIME: string;
    SUBJECT: string;
    DATACONTENTTYPE: string;
    DATASCHEMA: string;
}>
```

Otherwise, I'm not getting autocompletion for `KAFKA_CE_HEADERS.`...

File: src/message/kafka/index.ts [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+ * @param {CloudEvent<T>} event The event to serialize
+ * @returns {KafkaMessage<T>} a KafkaMessage instance
+ */
+function toBinaryKafkaMessage<T>(event: CloudEvent<T>): KafkaMessage<T> {
+  // 3.2.1. Content Type
+  // For the binary mode, the header content-type property MUST be mapped directly
+  // to the CloudEvents datacontenttype attribute.
+  const headers: Headers = { 
+    ...{ [CONSTANTS.HEADER_CONTENT_TYPE]: event.datacontenttype }, 
+    ...headersFor(event) };
```
Comment (grant) [inline src/message/kafka/index.ts:52]: nit: I'm surprised linting does not move `};` to a new line.

File: src/message/kafka/index.ts [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+    throw new ValidationError("Value is null or undefined");
+  }
+  if (!message.headers) {
+    throw new ValidationError("Headers are null or undefined");
+  }
+  const cleanHeaders: Headers = sanitize(message.headers);
+  const mode: Mode = getMode(cleanHeaders);
+  switch (mode) {
+    case Mode.BINARY:
+      return parseBinary(message as KafkaMessage<T>);
```
Comment (grant) [inline src/message/kafka/index.ts:108]: Type assertion `message as KafkaMessage<T>` can be done once (or fewer times) above.

File: src/message/kafka/index.ts [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+/**
+ * Creates a Kafka key for the event provided
+ * @param {CloudEvent} e an event
+ * @returns {string} the Kafka key
+ */
+ function createKey<T>(e: CloudEvent<T>): string {
+  return `${e.source}/${e.id}`;
+}
```
Comment (grant) [inline src/message/kafka/index.ts:139]: nit: indentation looks off in this section

File: test/integration/emitter_factory_test.ts [hunk 7]
```diff
@@ -37,11 +37,11 @@ export const fixture = new CloudEvent({
   [ext3Name]: ext3Value,
 });
 
-function axiosEmitter(message: Message, options?: Options): Promise<unknown> {
```
Comment (grant) [inline test/integration/emitter_factory_test.ts:40]: It should be a goal to not require parameterization for this file. Can we try adding this feature without a change to this file?

File: test/integration/emitter_singleton_test.ts [hunk 8]
```diff
@@ -12,36 +12,36 @@ import { fixture, assertStructured } from "./emitter_factory_test";
 import { rejects, doesNotReject } from "assert";
 
 describe("Emitter Singleton", () => {
```
Comment (grant) [inline test/integration/emitter_singleton_test.ts:14]: It should be a goal to not require parameterization for this file. Can we try adding this feature without a change to this file?

A separate test for parameterized messages would be fine if needed.

**Review by grant — CHANGES_REQUESTED** [review]
Few requested changes. Hope the comments are helpful.

Nothing particularly blocking.

I'm not familiar with the goal of the SDKs in terms of supporting all protocols / transports. It be best to keep other transport logic to child helpers since a developer will usually only use the SDK for one protocol at a time.

**Comment by lance** [comment]
Thanks for the feedback @grant - I've made some updates. Making the generic type `T` default to `string` was quite helpful - thanks for pointing that out.

**Comment by lance** [comment]
> I'm not familiar with the goal of the SDKs in terms of supporting all protocols / transports. It be best to keep other transport logic to child helpers since a developer will usually only use the SDK for one protocol at a time.

Not sure how I feel about this. Other SDKs do support multiple protocols, and all do so within a single repository. Distributing the bits of these implementations is, of course, language specific. So, in Java, for example, you can pull in just the Kafka bindings iiuc, but that's not the case in Rust, for example, where the implementation  is all together. In any case, the different transport protocol support is all implemented in a single github repo - which is what we would need to do as well. I think the additional bytes that are added to the package by these two files are really not an issue compared to the complexity of reorganizing this repo to be a mono-repo, supporting multiple npm modules. My 2c.

File: src/message/kafka/index.ts (hunk 5 above)
Comment (lance) [reply src/message/kafka/index.ts:109]: I tried to simplify this, but didn't have much luck at an elegant solution. If you can suggest a specific change for this that compiles I'd be happy to use it.

**Comment by grant** [comment]
> > I'm not familiar with the goal of the SDKs in terms of supporting all protocols / transports. It be best to keep other transport logic to child helpers since a developer will usually only use the SDK for one protocol at a time.
> 
> Not sure how I feel about this. Other SDKs do support multiple protocols, and all do so within a single repository. Distributing the bits of these implementations is, of course, language specific. So, in Java, for example, you can pull in just the Kafka bindings iiuc, but that's not the case in Rust, for example, where the implementation is all together. In any case, the different transport protocol support is all implemented in a single github repo - which is what we would need to do as well. I think the additional bytes that are added to the package by these two files are really not an issue compared to the complexity of reorganizing this repo to be a mono-repo, supporting multiple npm modules. My 2c.

Yeah it's fine imo. The main thing is being wary of extra dependencies that may be unnecessary, and ensuring it's still easy to contribute to HTTP features without having to affect other protocols.

File: src/message/kafka/index.ts [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+    headers[CONSTANTS.HEADER_CONTENT_TYPE]?.startsWith(CONSTANTS.MIME_CE_BATCH) as boolean; // A batch of events
+}
+
+/**
+ * Creates a Kafka key for the event provided
+ * @param {CloudEvent} e an event
+ * @returns {string} the Kafka key
+ */
+function createKey<T>(e: CloudEvent<T>): string {
+  return `${e.source}/${e.id}`;
```
Comment (LeoVS09) [inline src/message/kafka/index.ts:139]: I am glad you used my implementation, but I initially incorrectly implemented the specification.
`Key` cannot be constructed that way. Correct implementation require to use [Partition Key Extenstion](https://github.com/cloudevents/spec/blob/v1.0.1/kafka-protocol-binding.md#31-key-mapping).
I added that extension in [this commit](https://github.com/LeoVS09/cloudevents-js-sdk-kafka/commit/aed9bb89b21b59fa023ef94d94f6054d6004c622)

**It is important to not construct** `key` in such a way not only because of specification but also because this key will be used for partitioning by Kafka. It must be **not unique**. Different events, the order of which is important, must share the same `key`, and still have a unique `id` for correctly processing the Exactly-Once Delivery flow. 
**In simple words:** better to keep this field empty than try to construct it.

**Review by LeoVS09 — CHANGES_REQUESTED** [review]
Sorry for breaking into your working process, but it critical bug was my failure, and it can create complex problems for production applications

File: src/message/kafka/index.ts (hunk 9 above)
Comment (lance) [reply src/message/kafka/index.ts:139]: @LeoVS09 thank you very much for the feedback. I will make the necessary changes for this in the new year. Your implementation has been a very helpful inspiration for this addition and I appreciate both the time you've committed to the code, and that which you spent reviewing the pr. Thanks!

**Comment by lance** [comment]
@LeoVS09 I've made the appropriate changes, ptal.
@grant @lholmquist a bit has changed sine your last reviews, would you mind taking another look at this? Thanks.

**Comment by lance** [comment]
Marking as WIP until https://github.com/cloudevents/sdk-javascript/pull/457 is resolved as this PR depends on many of those changes, and I think it's better to keep the PRs small and focused vs. cramming all of that in here.

**Comment by lance** [comment]
I have rebased with the latest changes from `main` and generally cleaned this up a bit. It's ready for a final review @grant and/or @lholmquist. Thanks!

**Review by grant — APPROVED** [review]
